sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/caches/AsyncCacheNonBlocking.java

```diff
@@ -0,0 +1,208 @@
+     * </para>
+     * <para>
+     * If previous initialization function is successfully completed it will return the value. It is possible this
+     * value is stale and will only be updated after the force refresh task is complete.
+     * </para>
+     * <para>
+     * If previous initialization function failed - new one will be launched.
+     * </para>
+     * </summary>
+     */
+    public Mono<TValue> getAsync(
+        TKey key,
+        Function<TValue, Mono<TValue>> singleValueInitFunc,
+        boolean forceRefresh) {
+
+        AsyncLazyWithRefresh<TValue> initialLazyValue = values.get(key);
+        if (initialLazyValue != null) {
+            logger.debug("cache[{}] exists", key);
+            return initialLazyValue.getValueAsync().flatMap(value -> {
+                if(!forceRefresh) {
+                    return Mono.just(value);
+                }
+
+                Mono<TValue> refreshMono = initialLazyValue.createAndWaitForBackgroundRefreshTaskAsync(singleValueInitFunc);
+                AsyncLazyWithRefresh<TValue> asyncLazyWithRefresh = new AsyncLazyWithRefresh<TValue>(refreshMono);
+                this.values.put(key, asyncLazyWithRefresh);
+                AsyncLazyWithRefresh<TValue> result = this.values.get(key);
+
+                return result.getValueAsync().onErrorResume(
+                    (error) -> {
+                        logger.debug("refresh cache [{}] resulted in error", key, error);
+                        if (initialLazyValue.shouldRemoveFromCache()) {
+                            this.remove(key);
+                        }
+                        return Mono.empty();
+                    }
+                );
+            }).onErrorResume((error) -> {
+                logger.debug("cache[{}] resulted in error", key, error);
+                if (initialLazyValue.shouldRemoveFromCache()) {
+                    this.remove(key);
+                }
+                return Mono.empty();
+            });
+        }
+
+        logger.debug("cache[{}] doesn't exist, computing new value", key);
+        AsyncLazyWithRefresh<TValue> asyncLazyWithRefresh = new AsyncLazyWithRefresh<TValue>(singleValueInitFunc);
+        this.values.putIfAbsent(key, asyncLazyWithRefresh);
+        AsyncLazyWithRefresh<TValue> result = this.values.get(key);
+
+        return result.getValueAsync().onErrorResume(
+            (error) -> {
+                logger.debug("cache[{}] resulted in error", key, error);
+                // Remove the failed task from the dictionary so future requests can send other calls..
+                this.remove(key);
+                return Mono.empty();
+            }
+        );
+    }
+
+    public void set(TKey key, TValue value) {
+        logger.debug("set cache[{}]={}", key, value);
+        AsyncLazyWithRefresh<TValue> updatedValue = new AsyncLazyWithRefresh<TValue>(value);
+        this.values.put(key, updatedValue);
```

Comment:
usually a cache has a limited number of entries; should we consider the case where there might be too many entries and we need to "bound" the size of it?